### DISCUSSION TITLE
Rendering a '||' character with python '+' fails sometimes

### DISCUSSION THREAD

**Discussion by vinraspa** [sqlalchemy/sqlalchemy#9515] [Usage Questions] [answered]
First, I want to tell that I know I have to read documentation thoroughly before asking a question.
I have been trying to solve my problem for several days. I am posting here as a last intention.
Even ChatGPT could not help me while it generally does so I need an external natural intelligence to fill my own gaps ^^

Please, find below a MWE:
```
from dataclasses import dataclass
from datetime import date

from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
import sqlalchemy as SA


class Base(DeclarativeBase):
    pass


@dataclass
class FixedAsset(Base):
    __tablename__ = "Fixed_assets"
    id: Mapped[int] = mapped_column(primary_key=True)
    entry_date: Mapped[date] = mapped_column()
    duration: Mapped[int] = mapped_column()


print(SA.select(SA.func.date(FixedAsset.entry_date, "+" + FixedAsset.duration + " year", '-1 day')))
```
and the actual output:
```
SELECT date("Fixed_assets".entry_date, :duration_1 + "Fixed_assets".duration || :param_1, :date_2) AS date_1 
FROM "Fixed_assets"
```

My problem is that the _plus sign_ between `"+"` and `FixedAsset.duration` is rendered as is (a `+`) while the second _plus sign_ is rendered as the expected sqlite concat character `||`.

Two questions:

1. Why this difference in rendering;
2. How can I get the correct SQL statement, that is with two double-pipe charcaters `SELECT date("Fixed_assets".entry_date, :duration_1 || "Fixed_assets".duration || :param_1, :date_2) AS date_1`

In advance, thank you.

**Comment by zzzeek** [comment, marked as answer]
the behavior of the `+` operator switches on the kind of datatype that's in use.  it only produces `||`, the string concatenation operator, if you have string datatypes in place.

`FixedAsset.duration` appears to be an integer.  you can't concatenate integers to strings, you'd have to use SQL CAST to cast that integer as a string first:

    print(SA.select(SA.func.date(FixedAsset.entry_date, "+" + SA.cast(FixedAsset.duration, SA.String) + " year", '-1 day')))

**Reply by vinraspa** [reply]
So obvious! Thank you much!

**Reply by vinraspa** [reply]
Is there a way to render a `||` regardless of what is around?

**Reply by vinraspa** [reply]
I tried with `concat` but it produces a raw `concat` which is not understood by SQLite

**Reply by zzzeek** [reply]
there's `col.op('||')(other)`, but I think going with casts or type_coerce (which works also here) would be your best best

**Reply by vinraspa** [reply]
Fine! That's clear. Thanks and sorry for bothering.

**Reply by zzzeek** [reply]
totally fine questions!
